pkg/providers/oidc/provider.go

```diff
@@ -0,0 +1,79 @@
+package oidc
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/logger"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/requests"
+)
+
+// providerJSON resresents the information we need from an OIDC discovery
+type providerJSON struct {
+	Issuer      string `json:"issuer"`
+	AuthURL     string `json:"authorization_endpoint"`
+	TokenURL    string `json:"token_endpoint"`
+	JWKsURL     string `json:"jwks_uri"`
+	UserInfoURL string `json:"userinfo_endpoint"`
+}
+
+// Endpoints represents the endpoints discovered as part of the OIDC discovery process
+// that will be used by the authentication providers.
+type Endpoints struct {
+	AuthURL     string
+	TokenURL    string
+	JWKsURL     string
+	UserInfoURL string
+}
+
+// EndpointProvider allows the endpoints to be discovered
+type EndpointProvider interface {
+	Endpoints() Endpoints
+}
+
+// NewProvider allows a user to perform an OIDC discovery and returns the EndpointProvider.
+// We implement this here as opposed to using oidc.Provider so that we can override the Issuer verification check.
+// As we have our own verifier and fetch the userinfo separately, the rest of the oidc.Provider implementation is not
+// useful to us.
+func NewProvider(ctx context.Context, issuerURL string, skipIssuerVerification bool) (EndpointProvider, error) {
```

Comment:
Starting with PKCE, there is more information which we may want to pull out of this than just `Endpoints`. For instance in RFC-8414 implemented by Google, Dex, and Auth0 (to name a few), there is now the code challenge methods that are supported in the payload.

It would be great to see a way to pull additional claims out of this response. We currently were able to do that in `providers` by using the `oidcProvider.Claims(struct)` to map extra claims to a struct.

Reply:
Good point! Perhaps instead of this being an `EndpointsProvider` interface it should be a `DiscoveryProvider` and then we can extend the interface to include a new `PKCE` method to retrieve the PKCE information, WDYT?